File: lib/cells.js
'use strict';

const PREFIXES = require('../data/projects').map(p => p.prefix.toLowerCase());
const BROKEN = Object.entries(require('../data/broken')).reduce((o, [k, v]) => {
	o[k.toLowerCase()] = v;
	return o;
}, {});
const { isInterior, getCellName, CELL_SIZE, getCellGrid } = require('./util');
const MIN_Z = -32000;
const MAX_Z = 64000;

function isInvalidCoord(coord) {
	return !Number.isFinite(coord) || coord > Number.MAX_SAFE_INTEGER || coord < Number.MIN_SAFE_INTEGER;
}

function getPointCoords(point, record) {
	if(record.data?.grid) {
		const [x, y] = record.data.grid;
		if(x || y) {
			let [xPos, yPos] = point.location;
			xPos += x * CELL_SIZE;
			yPos += y * CELL_SIZE;
			return [point.location, `(${xPos}, ${yPos})`];
		}
	}
	return [point.location];
}

module.exports = {
	onRecord(record, mode, id) {
		if('deleted' in record) {
			return;
		}
		if(record.type === 'PathGrid') {
			const { points, connections } = record;
			if(!points?.length) {
				return;
			}
			const connected = new Set(connections);
			for(let i = 0; i < points.length; i++) {
				const point = points[i];
				if(point.connection_count) {
					connected.add(i);
				}
				for(let j = i + 1; j < points.length; j++) {
					const otherPoint = points[j];
					if(point.location.every((l, index) => l === otherPoint.location[index])) {
						console.log(record.type, getCellName(record), 'contains duplicate node at', ...getPointCoords(point, record));
						break;
					}
				}
			}
			if(points.length !== connected.size) {
				for(let i = 0; i < points.length; i++) {
					if(!connected.has(i)) {
						console.log(record.type, getCellName(record), 'contains unconnected node at', ...getPointCoords(points[i], record));
					}
				}
			}
		} else if(isInterior(record) && !record.atmosphere_data?.fog_density && !PREFIXES.some(prefix => id.startsWith(prefix))) {
			console.log(record.type, record.id, 'has a fog density of 0');
		}
	},
	onCellRef(record, reference, id) {
		if(!('deleted' in reference) && !isInterior(record)) {
			const [x, y] = record.data.grid;
			const xBound = CELL_SIZE * x;
			const yBound = CELL_SIZE * y;
			const [xPos, yPos, zPos] = reference.translation;
			if(reference.translation.some(isInvalidCoord) || zPos <= MIN_Z || zPos >= MAX_Z) {
				console.log(record.type, getCellName(record), 'contains far out reference', reference.id, 'at', reference.translation);
			} else if(xPos < xBound || yPos < yBound || xPos >= xBound + CELL_SIZE || yPos >= yBound + CELL_SIZE) {
				console.log(record.type, getCellName(record), 'contains out of bounds reference', reference.id, 'at', reference.translation, 'which should be in', getCellGrid(xPos, yPos));
			}
		}
		if(id in BROKEN) {
			const replacement = BROKEN[id];
			console.log(record.type, getCellName(record), 'contains broken reference', reference.id, replacement ? `which should be ${replacement}` : '');
		}
	}
};

File: src/validators/cells.rs
use super::Context;
use crate::{
    handlers::Handler,
    util::{get_cell_grid, CELL_SIZE},
};
use std::collections::{HashMap, HashSet};
use tes3::esp::{Cell, CellFlags, EditorId, PathGrid, PathGridPoint, Reference, TES3Object};

include!(concat!(env!("OUT_DIR"), "/gen_broken.rs"));

const MAX_Z: f32 = 64000.;
const MIN_Z: f32 = -32000.;
const MAX_SAFE_INT: f32 = 9007199254740991.;
const MIN_SAFE_INT: f32 = -9007199254740991.;
const BLACK_SQUARES: [&str; 4] = [
    "in_lava_blacksquare",
    "t_aid_blackcircle_01",
    "t_aid_blackcircle_02",
    "t_aid_blacktriangle_01",
];

pub struct CellValidator {
    seen: HashSet<String>,
    broken: HashMap<&'static str, &'static str>,
}

fn get_point_coords(point: &PathGridPoint, record: &PathGrid) -> String {
    let [x_pos, y_pos, _] = point.location;
    let location = format!("[{}, {}]", x_pos, y_pos);
    let (x, y) = record.data.grid;
    if x != 0 || y != 0 {
        let ext_x = x_pos + x * CELL_SIZE as i32;
        let ext_y = y_pos + y * CELL_SIZE as i32;
        return format!("{} ({}, {})", location, ext_x, ext_y);
    }
    location
}

fn get_water_height(cell: &Cell) -> Option<f32> {
    if cell.is_exterior() {
        return Some(-1.);
    } else if cell.data.flags.contains(CellFlags::HAS_WATER) {
        return cell.water_height;
    }
    None
}

impl Handler<'_> for CellValidator {
    fn on_record(&mut self, context: &Context, record: &TES3Object, _: &str, _: &str) {
        match record {
            TES3Object::Cell(cell) => {
                if cell.is_interior()
                    && cell
                        .atmosphere_data
                        .as_ref()
                        .is_some_and(|d| d.fog_density == 0.0)
                    && !context.projects.iter().any(|p| p.matches(&cell.name))
                {
                    println!("Cell {} has a fog density of 0", cell.editor_id());
                }
            }
            TES3Object::PathGrid(pathgrid) => {
                let points = &pathgrid.points;
                if points.is_empty() {
                    return;
                }
                let mut connected: HashSet<u32> = HashSet::new();
                connected.extend(&pathgrid.connections);
                for (i, point) in points.iter().enumerate() {
                    if point.connection_count > 0 {
                        connected.insert(i as u32);
                    }
                    for other_point in points[i + 1..].iter() {
                        if point
                            .location
                            .into_iter()
                            .enumerate()
                            .all(|(index, l)| l == other_point.location[index])
                        {
                            println!(
                                "PathGrid {} contains duplicate node at {}",
                                pathgrid.editor_id(),
                                get_point_coords(point, pathgrid)
                            );
                            break;
                        }
                    }
                }
                if points.len() != connected.len() {
                    for (i, point) in points.iter().enumerate() {
                        if !connected.contains(&(i as u32)) {
                            println!(
                                "PathGrid {} contains unconnected node at {}",
                                pathgrid.editor_id(),
                                get_point_coords(point, pathgrid)
                            );
                        }
                    }
                }
            }
            _ => {}
        }
    }

    fn on_cellref(
        &mut self,
        _: &Context,
        record: &Cell,
        reference: &Reference,
        id: &str,
        _: &[&Reference],
        _: usize,
    ) {
        if !reference.deleted.unwrap_or(false) && !record.is_interior() {
            let (x, y) = record.data.grid;
            let x_bound = CELL_SIZE * x as f64;
            let y_bound = CELL_SIZE * y as f64;
            let x_pos = reference.translation[0] as f64;
            let y_pos = reference.translation[1] as f64;
            let z_pos = reference.translation[2];
            if reference
                .translation
                .iter()
                .any(|coord| !coord.is_finite() || *coord > MAX_SAFE_INT || *coord < MIN_SAFE_INT)
                || !(MIN_Z..=MAX_Z).contains(&z_pos)
            {
                println!(
                    "Cell {} contains far out reference {} at [{}, {}, {}]",
                    record.editor_id(),
                    reference.id,
                    x_pos,
                    y_pos,
                    z_pos
                );
            } else if x_pos < x_bound
                || y_pos < y_bound
                || x_pos >= x_bound + CELL_SIZE
                || y_pos >= y_bound + CELL_SIZE
            {
                let (actual_x, actual_y) = get_cell_grid(x_pos, y_pos);
                println!(
                    "Cell {} contains out of bounds reference {} \
                at [{}, {}, {}] which should be in ({}, {})",
                    record.editor_id(),
                    reference.id,
                    x_pos,
                    y_pos,
                    z_pos,
                    actual_x,
                    actual_y
                );
            }
        }
        if let Some(replacement) = self.broken.get(&id) {
            if replacement.is_empty() {
                println!(
                    "Cell {} contains broken reference {}",
                    record.editor_id(),
                    reference.id
                );
            } else {
                println!(
                    "Cell {} contains broken reference {} which should be {}",
                    record.editor_id(),
                    reference.id,
                    replacement
                );
            }
        }
        if let Some(height) = get_water_height(record) {
            let [x, y, _] = reference.rotation;
            if record.is_interior()
                && reference.translation[2] >= height
                && x == 0.
                && y == 0.
                && BLACK_SQUARES
                    .iter()
                    .any(|id| id.eq_ignore_ascii_case(&reference.id))
            {
                let name = record.editor_id();
                let key = format!("{}_{}", name, id);
                if self.seen.insert(key) {
                    println!(
                        "Cell {} contains above water black square {}",
                        name, reference.id
                    );
                }
            }
        }
    }
}

impl CellValidator {
    pub fn new() -> Self {
        Self {
            seen: HashSet::new(),
            broken: get_broken_data(),
        }
    }
}
